In [1]:
'''
Notebook by Alon Agmon.
Based on Elkan & Noto "Learning classifiers from only positive and unlabeled data." 
Proceeding of the 14th ACM SIGKDD international conference on Knowledge discovery and data mining. ACM, 2008.
Implementaion code by Alexandre Drouin (https://github.com/aldro61)
'''


import pandas as pd
import numpy as np
url = "http://archive.ics.uci.edu/ml/machine-learning-databases/00267/data_banknote_authentication.txt"
data = pd.read_csv(url, header=None)

### Load data set

#### The data set will contain 4 features and target var indicating whether the note is forged or authentic

In [2]:
print(data.shape)
data.head(10)

(1372, 5)


0       1        2        3  4
0  3.62160  8.6661 -2.80730 -0.44699  0
1  4.54590  8.1674 -2.45860 -1.46210  0
2  3.86600 -2.6383  1.92420  0.10645  0
3  3.45660  9.5228 -4.01120 -3.59440  0
4  0.32924 -4.4552  4.57180 -0.98880  0
5  4.36840  9.6718 -3.96060 -3.16250  0
6  3.59120  3.0129  0.72888  0.56421  0
7  2.09220 -6.8100  8.46360 -0.60216  0
8  3.20320  5.7588 -0.75345 -0.61251  0
9  1.53560  9.1772 -2.27180 -0.73535  0

#### Check how balanced the data set is in terms of postives and negatives

In [3]:
data.iloc[:, -1].value_counts()

0    762
1    610
Name: 4, dtype: int64

#### Train a classifier to create a baseline

In [4]:
from sklearn.model_selection import train_test_split

x_data = data.iloc[:,:-1]
y_data = data.iloc[:,-1]

x_train, x_test, y_train, y_test = train_test_split(x_data, y_data, test_size=0.2, random_state=7)

In [12]:
import xgboost as xgb

model = xgb.XGBClassifier()

model.fit(x_train, y_train)

[14:37:18] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.4.0/src/learner.cc:1095: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


C:\Users\lida\Anaconda3\envs\ch\lib\site-packages\xgboost\sklearn.py:1146: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.300000012, max_delta_step=0, max_depth=6,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=100, n_jobs=4, num_parallel_tree=1, random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=1,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [13]:
y_predict = model.predict(x_test)

#### Establish our baseline

In [14]:
from sklearn.metrics import recall_score, precision_score, roc_auc_score, accuracy_score, f1_score

def evaluate_results(y_test, y_predict):
    print('Classification results:')
    f1 = f1_score(y_test, y_predict)
    print("f1: %.2f%%" % (f1 * 100.0)) 
    roc = roc_auc_score(y_test, y_predict)
    print("roc: %.2f%%" % (roc * 100.0)) 
    rec = recall_score(y_test, y_predict, average='binary')
    print("recall: %.2f%%" % (rec * 100.0)) 
    prc = precision_score(y_test, y_predict, average='binary')
    print("precision: %.2f%%" % (prc * 100.0)) 

    
evaluate_results(y_test, y_predict)

Classification results:
f1: 99.57%
roc: 99.57%
recall: 99.15%
precision: 100.00%


### Test the PU learning approach

#### Keep aside 25% of the positives -- they will be the only labeled samples

In [6]:
mod_data = data.copy()
#get the indices of the positives samples
pos_ind = np.where(mod_data.iloc[:,-1].values == 1)[0]
#shuffle them
np.random.shuffle(pos_ind)
# leave just 25% of the positives marked
pos_sample_len = int(np.ceil(0.25 * len(pos_ind)))
print(f'Using {pos_sample_len}/{len(pos_ind)} as positives and unlabeling the rest')
pos_sample = pos_ind[:pos_sample_len]


Using 153/610 as positives and unlabeling the rest


#### Create the target col 'class_test' that will be 1 for postive and -1 for unlabebed 

In [7]:
mod_data['class_test'] = -1
mod_data.loc[pos_sample,'class_test'] = 1
print('target variable:\n', mod_data.iloc[:,-1].value_counts())

target variable:
 -1    1219
 1     153
Name: class_test, dtype: int64


#### We now have just 153 positive samples labeled as 1 in the 'class_test' col while the rest is unlabeled as -1. 
#### Recall that col 4 still holds the actual label 

In [8]:
mod_data.head(10)

0       1        2        3  4  class_test
0  3.62160  8.6661 -2.80730 -0.44699  0          -1
1  4.54590  8.1674 -2.45860 -1.46210  0          -1
2  3.86600 -2.6383  1.92420  0.10645  0          -1
3  3.45660  9.5228 -4.01120 -3.59440  0          -1
4  0.32924 -4.4552  4.57180 -0.98880  0          -1
5  4.36840  9.6718 -3.96060 -3.16250  0          -1
6  3.59120  3.0129  0.72888  0.56421  0          -1
7  2.09220 -6.8100  8.46360 -0.60216  0          -1
8  3.20320  5.7588 -0.75345 -0.61251  0          -1
9  1.53560  9.1772 -2.27180 -0.73535  0          -1

#### Remember that this data frame (x_data) includes the former target variable that we keep here just to compare the results
[:-2] is the original class label for positive and negative data
[:-1] is the new class for positive and unlabeled data

In [9]:
x_data = mod_data.iloc[:,:-2].values # just the X 
y_labeled = mod_data.iloc[:,-1].values # new class (just the P & U)
y_positive = mod_data.iloc[:,-2].values # original class

#### The training set will be divided into a fitting-set that will be used to fit the estimator in order to estimate P(s=1|X) and a held-out set of positive samples that will be used to estimate P(s=1|y=1)
   

In [24]:
def fit_PU_estimator(X,y, hold_out_ratio, estimator):
    
    # find the indices of the positive/labeled elements
    assert (type(y) == np.ndarray), "Must pass np.ndarray rather than list as y"
    positives = np.where(y == 1.)[0] 
    # hold_out_size = the *number* of positives/labeled samples 
    # that we will use later to estimate P(s=1|y=1)
    hold_out_size = int(np.ceil(len(positives) * hold_out_ratio))
    np.random.shuffle(positives)
    # hold_out = the *indices* of the positive elements 
    # that we will later use  to estimate P(s=1|y=1)
    hold_out = positives[:hold_out_size] 
    # the actual positive *elements* that we will keep aside
    X_hold_out = X[hold_out] 
    # remove the held out elements from X and y
    X = np.delete(X, hold_out,0) 
    y = np.delete(y, hold_out)
    # We fit the estimator on the unlabeled samples + (part of the) positive and labeled ones.
    # In order to estimate P(s=1|X) or  what is the probablity that an element is *labeled*
    estimator.fit(X, y)
    # We then use the estimator for prediction of the positive held-out set 
    # in order to estimate P(s=1|y=1)
    hold_out_predictions = estimator.predict_proba(X_hold_out)
    #take the probability that it is 1
    hold_out_predictions = hold_out_predictions[:,1]
    # save the mean probability 
    c = np.mean(hold_out_predictions)
    return estimator, c

def predict_PU_prob(X, estimator, prob_s1y1):
    predicted_s = estimator.predict_proba(X)
    predicted_s = predicted_s[:,1]
    return predicted_s / prob_s1y1

#### test the PU estimation approach

In [25]:
predicted = np.zeros(len(x_data))
learning_iterations = 24
for index in range(learning_iterations):
    pu_estimator, probs1y1 = fit_PU_estimator(x_data, y_labeled, 0.2, xgb.XGBClassifier())
    predicted += predict_PU_prob(x_data, pu_estimator, probs1y1)
    if(index%4 == 0): 
        print(f'Learning Iteration::{index}/{learning_iterations} => P(s=1|y=1)={round(probs1y1,2)}')

[14:46:23] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.4.0/src/learner.cc:1095: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
Learning Iteration::0/24 => P(s=1|y=1)=0.11999999731779099
[14:46:23] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.4.0/src/learner.cc:1095: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[14:46:24] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.4.0/src/learner.cc:1095: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[14:46:24] 

#### compare the performance of the predictions of the PU approacj (y_predict) with the actuall original classes (y_positive) that we have saved aside

In [26]:
y_predict = [1 if x > 0.5 else 0 for x in (predicted/learning_iterations)]
evaluate_results(y_positive, y_predict)

Classification results:
f1: 58.47%
roc: 70.66%
recall: 41.31%
precision: 100.00%


In [ ]:
x_data = mod_data.iloc[:,:-2].values # just the X 
y_labeled = mod_data.iloc[:,-1].values # new class (just the P & U)
y_positive = mod_data.iloc[:,-2].values # original class

### PU Learning

In [47]:
from sklearn.svm import SVC
from pulearn import ElkanotoPuClassifier, WeightedElkanotoPuClassifier

In [38]:
svc = SVC(C=10, kernel='rbf', gamma=0.4, probability=True)
pu_estimator = ElkanotoPuClassifier(estimator=svc, hold_out_ratio=0.2)
pu_estimator.fit(x_data, y_labeled)

ElkanotoPuClassifier(estimator=SVC(C=10, gamma=0.4, probability=True),
                     hold_out_ratio=0.2)

In [45]:
y_pred = pu_estimator.predict(x_data)

In [42]:
def count_classes(y):
    tot_num = len(y)
    for i in np.unique(y):
        y_i = y[y == i]
        print(f'  label = {i}, pixel number = {y_i.shape[0]}, percentage = {round(len(y_i)/tot_num*100, 2)}%')

In [43]:
count_classes(y_labeled)

  label = -1, pixel number = 1219, percentage = 88.85%
  label = 1, pixel number = 153, percentage = 11.15%


In [44]:
count_classes(y_positive)

  label = 0, pixel number = 762, percentage = 55.54%
  label = 1, pixel number = 610, percentage = 44.46%


In [46]:
count_classes(y_pred)

  label = -1.0, pixel number = 5, percentage = 0.36%
  label = 1.0, pixel number = 1367, percentage = 99.64%


In [48]:
pu_estimator = WeightedElkanotoPuClassifier(estimator=svc, labeled=10, unlabeled=20, hold_out_ratio=0.2)
pu_estimator.fit(x_data, y_labeled)

WeightedElkanotoPuClassifier(estimator=SVC(C=10, gamma=0.4, probability=True),
                             hold_out_ratio=0.2, labeled=10, unlabeled=20)

In [49]:
count_classes(pu_estimator.predict(x_data))

  label = 1.0, pixel number = 1372, percentage = 100.0%


### OC-SVM

In [50]:
from sklearn.svm import OneClassSVM

In [52]:
clf = OneClassSVM(gamma='auto').fit(x_data)

In [57]:
y_pred = clf.predict(x_data)
count_classes(y_pred)

  label = -1, pixel number = 687, percentage = 50.07%
  label = 1, pixel number = 685, percentage = 49.93%


In [54]:
from sklearn.metrics import classification_report

In [58]:
y_pred[y_pred==-1] = 0
count_classes(y_pred)

  label = 0, pixel number = 687, percentage = 50.07%
  label = 1, pixel number = 685, percentage = 49.93%


In [61]:
print(classification_report(y_positive, y_pred))

              precision    recall  f1-score   support

           0       0.54      0.49      0.51       762
           1       0.43      0.49      0.46       610

    accuracy                           0.49      1372
   macro avg       0.49      0.49      0.49      1372
weighted avg       0.49      0.49      0.49      1372



In [62]:
from sklearn.metrics import confusion_matrix

In [63]:
confusion_matrix(y_positive, y_pred)

array([[373, 389],
       [314, 296]], dtype=int64)

In [64]:
"1010-101010_svm".split('_')[0]

'1010-101010'